src/mc-textopts.c

```diff
@@ -0,0 +1,305 @@
+        has_##Name = true;                                                                                             \
+    ((void)0)
+
+#define END_IF_FIELD                                                                                                   \
+    continue;                                                                                                          \
+    }                                                                                                                  \
+    else((void)0)
+
+#define ERROR_PREFIX "Error parsing TextOpts: "
+
+bool mc_TextOptsPerIndex_parse(mc_TextOptsPerIndex_t *txio, bson_iter_t *iter, mongocrypt_status_t *status) {
+    *txio = (mc_TextOptsPerIndex_t){0};
+    txio->set = true;
+
+    bool has_strMaxLength = false, has_strMinQueryLength = false, has_strMaxQueryLength = false;
+    while (bson_iter_next(iter)) {
+        const char *field = bson_iter_key(iter);
+        BSON_ASSERT(field);
+
+        IF_FIELD(strMaxLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxLength' must be greater than zero");
+                return false;
+            }
+            txio->strMaxLength = OPT_I32(val);
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(strMinQueryLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMinQueryLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMinQueryLength' must be greater than zero");
+                return false;
+            }
+            txio->strMinQueryLength = val;
+        }
+        END_IF_FIELD;
+
+        IF_FIELD(strMaxQueryLength);
+        {
+            if (!BSON_ITER_HOLDS_INT32(iter)) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxQueryLength' must be an int32");
+                return false;
+            }
+            const int32_t val = bson_iter_int32(iter);
+            if (val <= 0) {
+                CLIENT_ERR(ERROR_PREFIX "'strMaxQueryLength' must be greater than zero");
+                return false;
+            }
+            txio->strMaxQueryLength = val;
+        }
+        END_IF_FIELD;
+
+        CLIENT_ERR(ERROR_PREFIX "Unrecognized field: '%s'", field);
+        return false;
+    }
+    return true;
+}
+
+bool mc_TextOpts_parse(mc_TextOpts_t *txo, const bson_t *in, mongocrypt_status_t *status) {
+    bson_iter_t iter = {0};
+    BSON_ASSERT_PARAM(txo);
+    BSON_ASSERT_PARAM(in);
+    BSON_ASSERT(status || true);
+    bool has_caseSensitive = false, has_diacriticSensitive = false, has_substring = false, has_prefix = false,
```

Comment:
Suggest erroring if `caseSensitive` and `diacriticSensitive` options are not passed. The server [defaults to `true`](https://github.com/10gen/mongo/blob/5a2f1b9e213f4c7c1dde28faabffb486830f811d/src/mongo/crypto/encryption_fields.idl#L135-L144). I expect if the user's value does not match the server, that could lead to incorrect results rather than an error (similar to SERVER-91887).